datumaro/plugins/pointcloud_format/extractor.py

```diff
@@ -0,0 +1,134 @@
+
+        common_attrs = ["occluded"]
+        label_cat = LabelCategories(attributes=common_attrs)
+        if meta:
+            for label in meta["classes"]:
+                label_cat.add(label['title'])
+
+        categories[AnnotationType.label] = label_cat
+
+        data_dir = osp.join(osp.dirname(path), PointCloudPath.DEFAULT_DIR, PointCloudPath.ANNNOTATION_DIR)
+
+        labels = {}
+        for _, _, files in os.walk(data_dir):
+            for file in files:
+                with open(osp.join(data_dir, file), "r") as f:
+                    figure_data = json.loads(f.read())
+
+                for label in figure_data["objects"]:
+                    labels.update({label["key"]: label["classTitle"]})
+
+                group = 0
+                z_order = 0
+                attributes = {}
+
+                for figure in figure_data["figures"]:
+                    anno_points = []
+                    geometry_type = ["position", "rotation", "dimensions"]
+                    for geo in geometry_type:
+                        anno_points.extend(float(i) for i in figure["geometry"][geo].values())
+
+                    for _ in range(7):
+                        anno_points.append(0.0)
+
+                    map_id = mapping["figures"][figure['key']]
+                    label = labels[figure["objectKey"]]
+
+                    label = categories[AnnotationType.label].find(label)[0]
+
+                    shape = Cuboid(anno_points, label=label, z_order=z_order,
+                                   id=map_id, attributes=attributes, group=group)
+
+                    frame = mapping["videos"][figure_data['key']]
+                    frame_desc = items.get(frame, {'annotations': []})
+
+                    frame_desc['annotations'].append(shape)
+                    items[frame] = frame_desc
+
+        return items, categories
+
+    def _load_items(self, parsed):
+        for frame_id, item_desc in parsed.items():
+            name = item_desc.get('name', 'frame_%06d.png' % int(frame_id))
+            image = osp.join(self._images_dir, name)
+            image_size = (item_desc.get('height'), item_desc.get('width'))
+            if all(image_size):
+                image = Image(path=image, size=tuple(map(int, image_size)))
+
+            parsed[frame_id] = DatasetItem(id=osp.splitext(name)[0],
+                                           subset=self._subset, image=image,
+                                           annotations=item_desc.get('annotations'),
+                                           attributes={'frame': int(frame_id)})
+
+        return parsed
+
+
+class PointCloudImporter(Importer):
+    @classmethod
+    def find_sources(cls, path):
+        sources = cls._find_sources_recursive(path, '.json', 'point_cloud')
+        for i, source in enumerate(sources):
+            if osp.split(source["url"])[-1] == "meta.json" and source["format"] == 'point_cloud':
+                del sources[i]
```

Comment:
Don't change array/list size while iterating over it, because it can lead to wrong memory access. For resolution, create a copy and put all the necessary items there, for example.

Reply:
Resolved